packages/account/src/providers/provider.ts

```diff
@@ -495,10 +514,87 @@ export default class Provider {
         fullRequest = await options.requestMiddleware(fullRequest);
       }
 
-      return options.fetch ? options.fetch(url, fullRequest, options) : fetch(url, fullRequest);
+      const operationName = fullRequest.body
+        ?.toString()
+        .match(/"operationName":"(.+)"/)?.[1] as keyof SdkOperations;
+
+      if (BLOCK_HEIGHT_SENSITIVE_OPERATIONS.includes(operationName)) {
+        const currentBlockHeight = this.currentBlockHeightCache[url.replace(/-sub$/, '')] ?? 0;
+
+        fullRequest.body = fullRequest.body
+          ?.toString()
+          .replace(/}$/, `,"extensions":{"required_fuel_block_height":${currentBlockHeight}}}`);
+      }
+
+      return Provider.fetchAndProcessBlockHeight(url, fullRequest, options);
     }, retryOptions);
   }
 
+  private static async fetchAndProcessBlockHeight(
+    url: string,
+    request: RequestInit,
+    options: ProviderOptions
+  ): Promise<Response> {
+    let response: Response;
+    let blockHeightPreconditionFailed = false;
+
+    const retryOptions: RetryOptions = {
+      maxRetries: 5,
+      baseDelay: 500,
+    };
+
+    let retryAttempt = 0;
+
+    do {
+      response = await (options.fetch ? options.fetch(url, request, options) : fetch(url, request));
+
+      const responseClone = response.clone();
+
+      let extensions: {
+        current_fuel_block_height?: number;
+        fuel_block_height_precondition_failed: boolean;
+      };
+
+      if (url.endsWith('-sub')) {
+        const reader = responseClone.body?.getReader() as ReadableStreamDefaultReader<Uint8Array>;
+        const { event } = await FuelGraphqlSubscriber.readEvent(reader);
+
+        extensions = event?.extensions as typeof extensions;
+      } else {
+        extensions = (await responseClone.json()).extensions;
+      }
+
+      Provider.setCurrentBlockHeight(
+        url.replace(/-sub$/, ''),
+        extensions?.current_fuel_block_height
+      );
+
+      blockHeightPreconditionFailed = !!extensions?.fuel_block_height_precondition_failed;
+
+      if (blockHeightPreconditionFailed && retryAttempt < retryOptions.maxRetries) {
+        ++retryAttempt;
+        const sleepTime = getWaitDelay(retryOptions, retryAttempt);
+        await sleep(sleepTime);
+      } else {
+        break;
+      }
+    } while (blockHeightPreconditionFailed);
+
+    return response;
+  }
```

Comment:
Maybe I am missing something here, but why not use the variable to control the loop to consider `maxRetries` instead of explicitly breaking the loop?

Reply:
https://github.com/FuelLabs/fuels-ts/pull/3741/commits/cac250db6ca42f70e62435f98f919dca78bcddce